### ISSUE TITLE
Document gas cost estimates based on granularity of allowed personal Valuations

### ISSUE THREAD

**Issue by joshuahannan** [TrueBitFoundation/interactive-coin-offerings#7]
In the current implementation, we only allow bids to submit personal valuations in multiples of 100ETH.  There are a few reasons we think this is a good decision:

- In our linked list implementation, there can only be one node per valuation. So if there are two bids with the same valuation, there will be only one entry in the linked list for that valuation.  If we allow bidders to submit any valuation they want, there could potentially be as many nodes in the list as users participating in the sale. Since the list is sorted, finding the spot in the list to place a new valuation could potentially take a lot of operations which cost a lot of gas.  If we limit the granularity of personal valuations, there will be fewer nodes in the list, so fewer search operations, fewer gas.

- Bidders most likely will be fine if the final valuation is within $30000 of their predicted valuation, so there probably not much reason to allow them to submit more fine grain bids.

But there also might be some drawbacks:

-If there are fewer valuations to bid at, there is the chance that there will be a lot of bidders at the same valuation.  When auto-removing bids, doing this calculation to iterate through each bid, plus refunding fractions of bids for addresses at the same bid might end up being complex and expensive.
- bidders might actually want more flexibility with their valuations

Basically, we need someone to do some more quantitative research on relative gas costs for haveing a granularity of 100 ETH vs "infinite" granularity.  Gas costs for iterating through the linked list, calculating refunds for multiple addresses at the same valuation, etc.  We'll then be able to make a more education decision about implementation

**Comment by Shrugs** [comment]
Likewise about striking a balance. I agree re: +-$30k. Each project should be able to set that "uncertainty" themselves, of course, but I think that's a fine tradeoff, and after we do gas estimations each project will be able to make an informed decision.

**Comment by joshuahannan** [comment]
thats a good point. I think I'll actually make that a storage variable in the library that the owners of the sale can set themselves and then we'll just make a recommendation that owners limit the granularity to save the bidder's gas.  It'll still be useful to know the relative gas prices though for this and for the issue about when to perform the auto withdrawals.

**Comment by Hackdom** [comment]
So this will definitely need to be researched in this case because there is a limit to the number of nodes a bidder can traverse on chain. If the granularity is set such that there are too many nodes, not only is it more gas for the bidder, but the cost will exceed the block gas limit and effectively shut the auction down.

**Comment by Hackdom** [comment]
Updated the implementation using the three most significant value calculation. What this does is limit buckets to a logarithmic scale. When a cap is submitted, it can only define the first 3 values and then it must be trailing zeros after. ie You can submit a 101 Wei cap, but not a 1001 Wei cap. It must be 1010, 1100, etc. You can submit 156000000000000000000 as a cap but not 156100000000000000000. This solves the issue of too many buckets and doesn't rely on arbitrary input.

**Comment by teutsch** [comment]
Nice idea.
